lib/logstash/inputs/beats/message_listener.rb

```diff
@@ -32,11 +32,14 @@ def initialize(queue, input)
 
     def onNewMessage(ctx, message)
       hash = message.getData
-      ip_address = ip_address(ctx)
 
-      unless ip_address.nil? || hash['@metadata'].nil?
-        set_nested(hash, @input.field_hostip, ip_address)
+      if @input.include_source_metadata?
+        ip_address = ip_address(ctx)
+        unless ip_address.nil? || hash['@metadata'].nil?
+          set_nested(hash, @input.field_hostip, ip_address)
+        end
       end
+
       target_field = extract_target_field(hash)
```

Comment:
Why do we really need to override the codec? if we allow raw transfer when customer doesn't want codec metadata, it seems the problem will be solved!

```suggestion
      target_field = @input.include_codec_metadata?? extract_target_field(hash) : nil
```

Reply:
After stepping back and thinking, I think we are giving more flexibility for the users to control the metadatas rather than switching off (aligning on raw transformer) the codec.

Reply:
I understand your "multi-depth becomes the 1-depth" to be only a problem if:
(a) the structured event contains a top-level `message` _field_ with a _string_ value that holds a JSON-encoded payload
(b) the user instantiates this plugin with `codec => json`, causing the _decoded_ payload to get merged into the event as-sent by Beats

Because our _default_ codec is `plain`, and it creates an event from the `message` and the result is an event with a `message` that gets merged on-top of the existing `message` field, I don't see this to be practically a problem.

However, what you link to is likely a bug in which the `[event][original]` produced by the codec could _override_ an `[event][original]` that is part of the original beats payload :weary: